File: Dice throw.cpp
class Solution {
    static int noOfWays(int m, int n, int x) {
        // code here
        return rec(m,n,x,new HashMap<>());
    }
    private static int rec(int m,int n,int x,Map<String,Integer> hm){
        if(x==0){
            if(n==0) return 1;
            return 0;
        }
        String key=m+" "+n+" "+x;
        if(hm.containsKey(key)) return hm.get(key);
        int ways=0;
        for(int i=m;i>0;i--){
            if(x>=i) ways+=rec(m,n-1,x-i,hm);
        }
        hm.put(key,ways);
        return ways;
    }
};

File: Dice throw.py
class Solution:
    def noOfWays(self, m, n, x):
        dp = [[0] * (x + 1) for _ in range(n + 1)]
        dp[0][0] = 1

        for i in range(1, n + 1):
            for j in range(i, x + 1):
                for face in range(1, m + 1):
                    if j - face >= 0:
                        dp[i][j] += dp[i - 1][j - face]

        return dp[n][x]

